### PR TITLE
Add Route Select for Rate Limit

### PR DESCRIPTION
**Pull request by jsmolar**
This PR creates two new sets of objects:

- `RouteSelect`: https://docs.kuadrant.io/kuadrant-operator/doc/reference/route-selectors/#routeselector
- `HTTP Route Matchers`: https://gateway-api.sigs.k8s.io/reference/spec/#gateway.networking.k8s.io/v1.HTTPRouteMatch

The placement of these resources can be discussed as I was not 100% sure.
Tests will be added later to this PR.

Usage example:
```
   selector = RouteSelector(
        RouteMatch(
            path=HTTPPathMatch(type=MatchType.EXACT, value="/get"),
            headers=[HeadersMatch(name="header", value="value"), HeadersMatch(name="another", value="value")],
            query_params=[HTTPQueryParamsMatch(name="param", value="value")]
            method=HTTPMethod.GET,
        )
    )
    rate_limit.add_limit("multiple", [limit], route_selectors=[selector])
```

### CODE REVIEW

File: testsuite/policy/rate_limit_policy.py [hunk 1]
```diff
@@ -21,6 +21,17 @@ class Limit:
     unit: Literal["second", "minute", "day"] = "second"
 
 
+@dataclass
+class RouteSelect:
```
Comment (pehala) [inline testsuite/policy/rate_limit_policy.py:25]: `RouteSelector`

File: testsuite/gateway/__init__.py [hunk 2]
```diff
@@ -1,13 +1,15 @@
...
 from dataclasses import dataclass
-from typing import Any, Optional, TYPE_CHECKING
+from typing import Any, Optional, TYPE_CHECKING, Literal
 
 from httpx import Client
 
 from testsuite.certificates import Certificate
 from testsuite.lifecycle import LifecycleObject
-from testsuite.utils import asdict
+from testsuite.utils import asdict, _asdict_recurse
```
Comment (ficap) [inline testsuite/gateway/__init__.py:12]: In this situation I would consider making `_asdict_recurse` a public function as it seems that more and more code depends on it.

Comment (jsmolar) [reply testsuite/gateway/__init__.py:12]: Removed in last PR update

File: testsuite/utils.py [hunk 3]
```diff
@@ -138,9 +138,6 @@ def asdict(obj) -> dict[str, JSONValues]:
 
 
 def _asdict_recurse(obj):
-    if hasattr(obj, "asdict"):
```
Comment (pehala) [inline testsuite/utils.py:141]: I think changes to this  function can be reverted

Comment (jsmolar) [reply testsuite/utils.py:141]: removed

File: testsuite/policy/rate_limit_policy.py [hunk 4]
```diff
@@ -21,6 +21,23 @@ class Limit:
     unit: Literal["second", "minute", "day"] = "second"
 
 
+@dataclass
+class RouteSelector:
+    """
+    RouteSelector is an object composed of a set of HTTPRouteMatch objects (from Gateway API -
+    HTRTPPathMatch, HTTPHeaderMatch, HTTPQueryParamMatch, HTTPMethodMatch),
```
Comment (ficap) [inline testsuite/policy/rate_limit_policy.py:28]: typo: `HTRTPPathMatch` -> `HTTPPathMatch

File: testsuite/gateway/gateway_api/route.py [hunk 5]
```diff
@@ -73,17 +74,22 @@ def remove_all_hostnames(self):
...
-                    return
-        raise NameError("This backend is not assigned to this Route")
+    def add_rule(self, backend: "Httpbin", *route_matches: RouteMatch):
+        """Adds rule to the Route"""
+        rules = {"backendRefs": [backend.reference]}
+        matches = (
+            list(route_matches)
+            if route_matches
+            else [RouteMatch(path=PathMatch(type=MatchType.PATH_PREFIX, value="/"))]
+        )
```
Comment (pehala) [inline testsuite/gateway/gateway_api/route.py:84]:
```suggestion
   matches = list(route_matches)
   if len(matches) == 0:
       matches.append(RouteMatch(path=PathMatch(type=MatchType.PATH_PREFIX, value="/"))
```

File: testsuite/gateway/gateway_api/route.py [hunk 6]
```diff
@@ -73,17 +74,22 @@ def remove_all_hostnames(self):
...
+
+        rules.update({"matches": list(map(asdict, matches))})
```
Comment (pehala) [inline testsuite/gateway/gateway_api/route.py:86]:
```suggestion
        rules["matches"] = [asdict(match) for match in matches]
```

File: testsuite/gateway/__init__.py [hunk 7]
```diff
@@ -45,6 +46,69 @@ def reference(self) -> dict[str, Any]:
...
+
+    name: str
+    value: str
+    type: Optional[Literal[MatchType.EXACT, MatchType.REGULAR_EXPRESSION]] = None
+
+
+@dataclass
+class RouteMatch:
+    """
+    Abstract Dataclass for HTTPRouteMatch.
```
Comment (pehala) [inline testsuite/gateway/__init__.py:100]: Description no longer valid

File: testsuite/gateway/gateway_api/route.py (hunk 5 above)
Comment (jsmolar) [reply testsuite/gateway/gateway_api/route.py:84]: I somehow agree with your suggestion.
But the thing to discuss here is that if no `route_matches` is used for the rule,
```
        - path:
            type: PathPrefix
            value: /
```
is used automatically. No need to specifically defined `RouteMatch(path=PathMatch(type=MatchType.PATH_PREFIX, value="/"))`

Comment (pehala) [reply testsuite/gateway/gateway_api/route.py:84]: in that case, the condition does not need to be there

**Review by pehala — APPROVED** [review]

File: testsuite/gateway/__init__.py [hunk 8]
```diff
@@ -1,7 +1,8 @@
 """Classes related to Gateways"""
+import enum
 from abc import ABC, abstractmethod
 from dataclasses import dataclass
-from typing import Any, Optional, TYPE_CHECKING
+from typing import Any, Optional, TYPE_CHECKING, Literal, List
```
Comment (averevki) [inline testsuite/gateway/__init__.py:5]: We're moving to `list` from a standard python library. Not a PR blocker though

**Review by averevki — APPROVED** [review]
LGTM. Is this a setup for tests in the feature PR, or do you want to also add them here?

**Pull request merged**